### ISSUE TITLE
Support for bcrypt(md5($p)) on FPGA or --single && --stdout?

### ISSUE THREAD

**Issue by knewl551** [openwall/john#4261]
Not a bug but a feature request.

To audit bcrypt hashes, I use an old ZTEX clone board. Works perfect and fast compared to my 'conventional' GPU-rig (big thanks for supporting FPGAs!).

I want to audit a number of bcrypt(md5($p)) hashes now, at full ZTEX speed. The quick workaround is using --format=bcrypt-ztex --wordlist=md5_of_wordlist. However the powerful --single mode doest not work and the combination of --single and --stdout to do the same trick is not working ('Invalid options combination or duplicate option').

I understand that adding new FPGA algorithms is very complex. However, I thought that it might be doable to add format bcryptmd5-ztex where the onchanged bcrypt bitstream is used and MD5 pre and post processing is done by the CPU (since the CPU is by far fast enough to MD5 on the fly to keep the ZTEX busy).

Or more simple: is it possible to add support for --single in combination with --stdout so I can do the MD5 processing myself?

Or perhaps smarter tricks I didn't think of? :) Thank you.

**Comment by solardiz** [comment]
Yes, a bcryptmd5-ztex format like you describe is doable. However, we don't even have this kind of format on CPU, nor in OpenCL, so it'd end up being specialized to the specific combination of these two hashes and users of ZTEX boards. That's somewhat exotic. While I don't mind having a format like this, we don't intend to actually work on it unless you provide some incentive (pay for the effort? something else?) or implement this yourself and send us a pull request.

Also, I previously heard of uses of bcrypt on top of SHA-256 or SHA-512, in different encodings of those hashes (hex, base64). People do that to avoid bcrypt's length limit of 72. Maybe it'd make more sense (be less exotic) to add support for the whole set of those at once, rather than just the MD5. I suppose the combination with MD5 is a result of upgrading a system from MD5 to bcrypt (that's fine if so).

A question is how to recognize and how to record those hashes in `john.pot`. I suppose they look just like regular bcrypt hashes? Should we introduce our own prefixes indicating the inner hash type? Only for recording in `john.pot` or also for original expected input?

Regarding (not) supporting `--single --stdout`, this can't possibly work (except when run on just one password file line) without us inventing some format to encode the combinations of target hash (or target username?) and words to test against it. Historically, Cracker Jack had such a format - in fact, it couldn't generate and test the passwords at the same time, requiring that first a special-format wordlist be produced with one tool and then used to crack the hashes with another. I designed John's `--single` as an improvement upon that. Sure we could optionally revert to Jack's approach (by writing quite some extra code), but this seems rather specialized. Would there be many uses? I doubt it. Anyway, this would be best discussed on its own issue.

**Comment by solardiz** [comment]
Related: #4095.

For your own immediate use, I suppose it'd be the easiest to hack the existing bcrypt-ztex format turning it into what you need, by simply invoking MD5 and hex-encoding from within this format's `set_key`. You'd have this in a copy of the JtR tree that you'd use only for these hashes (as it'd have support for regular bcrypt hashes on ZTEX removed in favor of this flavor).

Edit: in a hack like the above, you'd also need to replace the test vectors with those corresponding to bcrypt(MD5).

**Comment by magnumripper** [comment]
@jfoug Perhaps some variant of dynamic compiler mode could be added for doing things like this generically and then send to chosen format?

**Comment by AlekseyCherepanov** [comment]
In CMIYC 2019, there was "Alaska" "task" where hashes were `django-scrypt(md5($p))` where `$p` is an old password for the same user, so I investigated ways to pass `md5($p)` as candidates for `--single`: [the final idea](https://www.openwall.com/lists/john-users/2019/10/09/1) is tricky, there are [more notes and a script to pack candidates together with hashes](https://www.openwall.com/lists/john-users/2019/10/10/2), and [the thread](https://www.openwall.com/lists/john-users/2019/10/06/1) describes why it is so tricky.
